File: src/lang/java/Fastpow.java

class Fastpow {

    public static long fastpow(long a, long b, long MOD) {
        long base = a % MOD, res = 1;
        while (b != 0) {
            if ((b & 1) == 1) {
                res = res * base % MOD;
            }
            base = base * base % MOD;
            b >>>= 1;
        }
        return res % MOD;
    }

    public static long fastpow(long a, long b) {
        int MOD = (int) 1e9 + 7;
        return fastpow(a, b, MOD);
    }
}

File: src/lang/typescript/fastpow.ts
export function fastPow(a: number, n: number): number {
    const $ = BigInt
    const MOD = $(1e9 + 7)
    let b = $(a) % MOD
    let res = 1n
    while (n) {
        if (n % 2 === 1) res = (res * b) % MOD

        b = (b * b) % MOD
        n = Math.trunc(n / 2)
    }
    return Number(res % MOD)
}
